### PR TITLE
fix(sources): promote profile samples by owning source, not computerId

### LINKED ISSUE #1149: setPrimaryDataSource strands a dive with no primary profile when the new source has no computerId
Found while investigating #623. Filed separately to keep that fix reviewable.

## Problem

`DiveRepositoryImpl.setPrimaryDataSource` demotes every profile row for the dive unconditionally, then re-promotes only rows matching the new primary's `computerId`, and only when that id is non-null:

```dart
// Demote all profiles for this dive.
await (_db.update(_db.diveProfiles)
      ..where((t) => t.diveId.equals(diveId)))
    .write(const DiveProfilesCompanion(isPrimary: Value(false)));

// Promote profiles belonging to the new primary computer.
if (newPrimary.computerId != null) {
  await (_db.update(_db.diveProfiles)..where(
        (t) => t.diveId.equals(diveId) &
               t.computerId.equals(newPrimary.computerId!),
      ))
      .write(const DiveProfilesCompanion(isPrimary: Value(true)));
}
```

When the newly chosen primary source has a null `computerId`, the promote is skipped entirely and the dive is left with **zero** rows where `is_primary = 1`.

## This is reachable, not theoretical

File-imported sources have a null `computerId` on both sides:

- The `dive_data_sources` row: `uddf_entity_importer.dart` sets `computerModel` and `computerSerial` but never `computerId`, so the column defaults to NULL. This is the funnel for the whole file-import pipeline.
- The profile rows: the `saveDive` batch insert in `dive_repository_impl.dart` never sets `computerId` either.

And the UI offers such a source as a valid choice. In `data_sources_section.dart` the "Set Primary" menu item is gated only on the source not already being primary:

```dart
if (!source.isPrimary && onSetPrimary != null)
  PopupMenuItem(value: _SourceMenuAction.setPrimary, ...)
```

There is no `computerId` guard anywhere on that path.

**Repro:** take a dive with two or more data sources, at least one of which came from a file import, then choose "Set Primary" on the file-imported source.

Two call sites reach the repository method:
- `lib/features/dive_log/presentation/pages/dive_detail_page.dart:5096` (the source picker)
- `lib/features/data_quality/data/services/quality_repair_executor.dart:325` (the data-quality repair action)

## Why it is easy to miss

The dive keeps rendering normally, because the display paths do not filter on `is_primary`: `getDiveById`, `getMergedProfile` and `getTimeAtDepthRanges` are all unfiltered. Only the `is_primary = 1` consumers silently skip the dive, among them:

- `getAscentDescentRates` (the Statistics "Average Ascent and Descent Rates" card)
- `getDiveProfile`, which filters `isPrimary.equals(true)`

So the profile is visibly present on the dive page while a subset of aggregates quietly ignores it, with no error and no user-facing signal.

## Suggested fix

Promote by data source rather than by computer. Options, roughly in order of preference:

1. Give profile rows a `sourceId` FK to `dive_data_sources` and promote on that, which removes the null-`computerId` special case entirely. Larger change, but it is the modelling problem underneath this bug.
2. Failing that, add an `else` branch promoting rows where `computer_id IS NULL` when `newPrimary.computerId` is null. Correct for the single-file-source case, ambiguous when a dive has two distinct file-imported sources, since both their profile row sets carry a null `computerId` and are indistinguishable.
3. At minimum, assert post-condition: if the update leaves the dive with no `is_primary` row, roll the transaction back rather than committing the stranded state.

Add a regression test that sets a null-`computerId` source as primary and asserts the dive still has at least one primary profile row.

## Related

The two coexisting load behaviours for `dive_profiles`, filtered versus unfiltered, are the backdrop here and are already noted in the codebase around `getMergedProfile` and `getDiveProfile` (see #536).

## Severity

Higher than a display bug. It silently corrupts a queryable flag through a normal UI action, and the user gets no indication anything went wrong.

### PR DESCRIPTION
**Pull request by ericgriffin**
## Summary

Fixes #1149. `setPrimaryDataSource` demoted every `dive_profiles` row for a dive, then re-promoted only rows matching the new primary's `computerId`, and only inside an `if (computerId != null)` guard. File-imported sources carry a null `computerId` on both sides (the `dive_data_sources` row and the sample rows), so promoting one demoted everything and promoted nothing, leaving the dive with **zero** `is_primary` rows.

It is silent because the display paths do not filter on the flag. `getDiveById` and `getMergedProfile` keep rendering the profile while `getDiveProfile`, `getAscentDescentRates` and the data-quality prefilters skip the dive entirely, so a subset of aggregates quietly loses a dive whose samples are visibly still there. Two call sites reach it: the dive detail source picker and the data-quality "set primary source" repair.

This takes option 1 from the issue: give sample rows a real owner instead of re-deriving one from `computerId`.

## Changes

- **`dive_profiles.source_id`**: nullable FK to `dive_data_sources`, `ON DELETE SET NULL` so samples outlive the metadata row describing them. `computerId` cannot identify an owner on its own, and two file-imported sources on one dive are reachable today by consolidating two file imports (the same-computer guard compares computer ids, and two nulls pass it).
- **Migration v154**: PRAGMA-guarded `ADD COLUMN` plus a one-time backfill that reproduces exactly what `getProfilesByDataSource` already did, so no existing row changes owner as a result of upgrading. The backfill probes **both** tables' columns, because `PRAGMA table_info` returns empty for a missing table and migration fixtures routinely hold `dive_profiles` without `dive_data_sources`. `beforeOpen` re-asserts the column only, never the full-table pass.
- **`setPrimaryDataSource`**: resolves the promote set **before** demoting, and only demotes when that set is non-empty. The stranding was never limited to a null `computerId`: demote-then-promote meant promoting *any* source owning no sample rows stranded the dive, including a metadata-only snapshot with a good computer id. This makes the stranded state unreachable rather than merely detected, and no legitimate action fails with an exception.
- **`dive_split_service`**: carried the identical `promoted.computerId != null` gate on its own re-promote, with the same result on the dive left behind. Fixed with the same rule.
- **Edited profiles**: promoting a source wholesale would have traded stranding for a visible double, since `saveEditedProfile` does not replace the rows it supersedes but demotes them and inserts a second null-`computerId` generation in the same source. The promote set is deduplicated by timestamp preferring the null-`computerId` row, the same "the edit is the live one" rule `restoreOriginalProfile` encodes, with a deterministic tiebreak so peers do not disagree and flip the flag back and forth through sync.
- **Writers carry attribution**: the download path stamps the source it just wrote, `saveEditedProfile` claims the source primary at edit time, reparse rewrites in place under the same source, and consolidate/merge/split remap old source ids to the copies they mint. `dive_merge_service` now inserts its data-source rows ahead of the profile copy, because `source_id` is a real FK and `foreign_keys` is ON, so SQLite enforces the reference immediately rather than at COMMIT.
- **`getProfilesByDataSource`** prefers the FK, which also stops a dive carrying two file-imported sources from being misread as edited and silently dropping the second source's samples.
- **Sync**: no serializer work needed (`dive_profiles` round-trips through generated `toJson`/`fromJson`), but `source_id` is registered in `parentRefs` as nullable so a profile whose source row is absent has the reference cleared instead of dangling the deferred-FK COMMIT.

### Tests

Two existing tests asserted the bug as intended behaviour, each with a comment presenting the broken output as the design ("All profiles should be demoted when new primary has no matching computerId", "no profiles are re-promoted"). Both now assert the dive keeps a primary profile.

New coverage:
- promotion follows the FK when `computerId` is null on both sides
- a source owning no samples leaves the existing primary alone
- an edited profile survives a swap away and back without duplicating
- v154 adds the column, attributes rows the way the pre-v154 read path did, and re-runs idempotently

## Test Plan

- [x] `flutter test` passes — 18932 tests, 0 failures, exit 0
- [x] `flutter analyze` passes — no issues found, whole project
- [x] `dart format` clean
- [x] Manual testing: not yet exercised against a real multi-source library; the repro (a dive with a file-imported source plus one other, then "Set Primary" on the file-imported one) is covered by automated tests only

### Note for reviewers

An earlier local run reported a `sync_retirement_fence_test` failure that did **not** reproduce; two other test suites were running concurrently on the same machine at the time. The whole sync directory passes in isolation (666 tests), as does the clean full run above.

### CODE REVIEW

File: lib/features/dive_log/data/repositories/dive_repository_impl.dart [hunk 1]
```diff
@@ -5874,19 +5931,23 @@ class DiveRepository {
...
+        // dive kept rendering, because getDiveById and getMergedProfile do
+        // not filter on the flag, while getDiveProfile, getAscentDescentRates
+        // and the data-quality prefilters silently skipped it.
+        final promote = await _profileIdsOwnedBySource(diveId, newPrimary);
+        if (promote.isNotEmpty) {
+          await (_db.update(_db.diveProfiles)
+                ..where((t) => t.diveId.equals(diveId)))
+              .write(const DiveProfilesCompanion(isPrimary: Value(false)));
+          await (_db.update(_db.diveProfiles)..where((t) => t.id.isIn(promote)))
               .write(const DiveProfilesCompanion(isPrimary: Value(true)));
```
[the remark this answers is not in the document]
Comment (ericgriffin) [reply lib/features/dive_log/data/repositories/dive_repository_impl.dart]: Confirmed and fixed. The cap is real: `sync_data_serializer.dart:1923` already documents it and chunks at 900. I probed the bundled build to get the actual number, and it is 32766 (SQLite 3.53.3), so the threshold is higher than 999 here but build-dependent, and this app does not use one SQLite everywhere (bundled SQLCipher on desktop, system library on Android). A regression test that promotes a 32767-sample dive reproduced it exactly: `SqliteException(1): too many SQL variables`.

I went with the predicate alternative rather than chunking, since chunking still encodes an assumption about whichever ceiling the current build has. The promote now binds a fixed three variables at any dive length.

Worth flagging what that turned up, because the first predicate version was worse than the bug it fixed. Correlating on `p.timestamp = dive_profiles.timestamp` rescans the whole dive once per row, because `dive_profiles` is indexed on `dive_id` alone. On the 32767-sample dive that measured **45 seconds**, which in the app is the UI frozen for 45 seconds on one "Set Primary" tap. Replacing it with `ROW_NUMBER() OVER (PARTITION BY timestamp ORDER BY (computer_id IS NULL) DESC, id DESC)` sorts the dive's rows once instead, and the same test now runs in **2.6 seconds**. Window functions are already used for the dedupe migrations in `database.dart`.

One detail worth calling out in the new predicate: `computer_id IS ?` rather than `=`. `=` never matches NULL, which is the entire shape of #1149, so the null-safe operator is what lets a single predicate cover both a real computer id and the null every file import carries.

**Comment by ericgriffin** [comment]
Rebased onto `main` (79 commits) and addressed the review comment.

**Schema version renumbered v154 to v156.** The migration was v154 when this branch opened; `main` has since taken v154 (#1104, site entry/exit method) and v155 (#828, gas model). The conflict was confined to `database.dart` and was five instances of the same ladder collision: `currentSchemaVersion`, the `migrationVersions` entry, the `onUpgrade` block, the `beforeOpen` backstop, and the helper block. All four numbered sites moved together, and the migration test file was renamed to `migration_v156_profile_source_id_test.dart` with its fixtures rebased to `user_version = 155`.

Renumbering rather than keeping 154 is deliberate: a device already stamped 154 by #1104 would never re-enter `if (from < 154)` and would be left without `source_id`. The `beforeOpen` backstop would self-heal it, but leaning on the backstop is not a resolution when the next number is free.

`minimumCompatibleSchemaVersion` stays at 137. This adds a new nullable column, which the guidance on that constant explicitly lists under "Do NOT raise it".

Receipt that nothing of main's work was reverted in the resolution: `git diff origin/main -- lib/core/database/database.dart` contains only this feature's added lines plus the one-line version bump.

Note for whoever merges #1127 and #1158: both still claim versions main has taken (155 and 154 respectively), so they will need the same renumber. Nothing currently claims 157.

**Verification after rebase:** `flutter test` 19189 passed / 0 failed / exit 0; `flutter analyze` no issues, whole project; `dart format` clean.

File: lib/features/dive_log/data/services/dive_split_service.dart [hunk 2]
```diff
@@ -270,6 +271,10 @@ class DiveSplitService {
                 .copyWith(
                   id: Value(_uuid.v4()),
                   diveId: Value(newDiveId),
+                  // Follow the source to its row on the new dive (issue
+                  // #1149); the original row is deleted at step 8, and
+                  // ON DELETE SET NULL would otherwise strip attribution.
```
[the remark this answers is not in the document]
Comment (ericgriffin) [reply lib/features/dive_log/data/services/dive_split_service.dart:286]: Confirmed and fixed in dd81382. The failure was more severe than described: I reproduced it with a dive carrying two file-imported sources (A primary with demoted rows, B secondary) and splitting B left the original dive with **zero** profile rows. Source A's samples did not merely get re-attributed, they were copied to the new dive and then deleted from the original at step 8.

Every branch is now additionally gated on `t.sourceId.isNull() | t.sourceId.equals(source.id)`. That only ever narrows the selection: a row whose `sourceId` names another source was never this source's to move, and rows carrying no `sourceId` still fall through to the legacy rules unchanged, so pre-v156 data behaves exactly as before. All six existing split tests still pass, including the primary-split-away promotion path.

Same root cause as #1149 itself one method over, which is the useful part of the catch: identifying rows by `computerId` when the owning source is what actually answers the question.

Note I deliberately did not widen the selection to include FK-owned rows that the legacy rules currently exclude. The "a computer-less source moves only non-primary null rows, user-edited primary rows stay with the original dive" rule is pre-existing and intentional, and changing it is a separate behavioural decision from the data-loss bug you flagged.

File: lib/features/dive_log/data/repositories/dive_repository_impl.dart [hunk 3]
```diff
@@ -6011,6 +6070,101 @@ class DiveRepository {
...
+      '(p.source_id = ?2 OR (p.source_id IS NULL AND p.computer_id IS ?3))';
+
+  List<Variable<Object>> _ownershipVars(
+    String diveId,
+    DiveDataSourcesData source,
+  ) => [
+    Variable<String>(diveId),
+    Variable<String>(source.id),
+    Variable<String>(source.computerId),
+  ];
```
[the remark this answers is not in the document]
Comment (ericgriffin) [reply lib/features/dive_log/data/repositories/dive_repository_impl.dart:6193]: Checked this one and I do not think it holds. Leaving the code as is, with the evidence.

Drift declares `final class Variable<T extends Object> extends Expression<T>` with `final T? value` (`drift-2.34.3/lib/src/runtime/query_builder/expressions/variables.dart:8-10`). The value field is explicitly nullable, so `Variable<String>(source.computerId)` with a `String?` argument is the correct spelling and type-checks. `flutter analyze` is clean across the project.

The suggested alternative does not compile. `Variable<String?>` is rejected outright:

```
error • 'String?' doesn't conform to the bound 'Object' of the type parameter 'T' • type_argument_not_matching_bounds
```

So the nullable type argument is not expressible in drift's API; the nullability lives in the field, not the type parameter.

NULL also binds as genuine SQL NULL rather than degrading to a string. Probed directly against the app database:

```dart
db.customSelect("SELECT (NULL IS ?1) AS null_matches, typeof(?1) AS t",
    variables: [Variable<String>(null)])
// null_matches=1  typeof=null
```

That is also covered indirectly by a test already on the branch: "keeps unattributed null-computerId profiles primary across a swap" exercises the legacy path where both the source and the rows have a null `computerId`, so it passes only if `p.computer_id IS ?3` matches NULL. If the bind were wrong, `_sourceOwnsProfiles` would return false, the promote would be skipped, and that test would fail.

Happy to revisit if you have a case where it misbehaves that these do not cover.

**Comment by ericgriffin** [comment]
Rebased onto `main` again (70 commits). Three conflicts this time, two of them real code rather than the ladder.

**Schema version v156 to v158.** `main` took 156 (#1127 travel gas) and 157 (#829 service price) while this sat open, so this is the second renumber (154 to 156 to 158). Same four-site move plus the test file, now `migration_v158_profile_source_id_test.dart` with fixtures rebased to `user_version = 157`. `minimumCompatibleSchemaVersion` stays at 137: still a new nullable column. Receipt holds, `git diff origin/main -- lib/core/database/database.dart` shows only this feature's additions plus the one-line version bump, and main's five v156 travel-gas references are untouched.

**`reparse_service.dart`** conflicted because #1164 added an `ownsStrand` guard around the same call this branch added a `sourceId` argument to. Kept both: the guard wraps the call, the call carries the argument. The diff against main is now purely additive.

**`dive_merge_service.dart`** conflicted because this branch moves the data-source insert loop ahead of the profile copy (`sourceId` is a real FK and `foreign_keys` is ON, so parents must land before children), while main grew a substantial comment block on that loop covering #1045, #1005 and #1164. The loop and all of main's commentary moved together to the earlier position; I verified every claim in it survives, and the step-10 slot now points at the new location.

Two things worth surfacing from that comment, since they intersect this PR directly:

- It states the underlying gap as "dive_profiles attributes samples by computerId alone, so computerId gets treated as a per-dive unique source key the schema never guaranteed, fixing that needs a `dive_profiles.data_source_id` FK, not a lossy write here." That FK is what this PR adds, so I updated that closing sentence to say the gap is now closed by `dive_profiles.sourceId` and each carried row keeps its own id with the copied samples pointing at it.
- Its `CAVEAT (#1164)` says "ReparseService step 4 calls `_replaceDiveProfiles` unconditionally, so re-parsing a merged dive today still wipes the merged profile." That now looks stale on main independently of this PR: `_sourceOwnsProfileStrand` returns false when no row on the dive is primary, which is exactly the merged-dive shape, so the call is already guarded. I carried the caveat over verbatim rather than editing someone else's comment inside a conflict resolution, but it probably wants deleting by whoever owns #1164.

Also worth noting for later: `_sourceOwnsProfileStrand` decides strand ownership by comparing `computerId` with an explicit "null counts as equal to null" rule, which is the same shape of question `sourceId` can now answer directly. Not touching it here, just flagging it as a natural follow-up.

**Verification:** `flutter analyze` no issues whole project; `dart format` clean; `flutter test` 19398 tests with a single failure in `media_item_view_store_fallback_test`, which is unrelated to this branch. It passes in isolation (9/9) and with the whole media directory (1497), this diff touches no media code, and the previous full run on this branch was 19189/0 including media. It appeared only after pulling main's 70 commits, so it looks like a full-suite parallelism flake rather than anything here. I have not run main's suite standalone to confirm it pre-exists.
